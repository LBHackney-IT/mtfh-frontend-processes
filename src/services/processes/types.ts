export enum Trigger {
  CloseProcess = "CloseProcess",
  CancelProcess = "CancelProcess",
}
