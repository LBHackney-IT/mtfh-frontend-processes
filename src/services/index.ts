export { default as locale } from "./locale";
export { default as menu } from "./menu";
export * as processes from "./processes";
export * from "./config";
