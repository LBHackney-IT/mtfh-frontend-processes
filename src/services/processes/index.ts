export * from "./sole-to-joint";
export * from "./change-of-name";
