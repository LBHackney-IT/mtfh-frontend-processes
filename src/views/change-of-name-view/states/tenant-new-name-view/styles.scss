@import '~lbh-frontend/lbh/base';
@import '~govuk-frontend/govuk/helpers/media-queries';

.mtfh-person-form {
  @include govuk-media-query($from: tablet) {
    max-width: 458px;
  }

  &__input {
    &--full {
      width: 100%;
    }

    &--contained {
      width: 300px;
      max-width: 100%;
    }
  }
}
