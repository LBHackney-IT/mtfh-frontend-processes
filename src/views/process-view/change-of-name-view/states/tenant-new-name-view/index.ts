export * from "./tenant-new-name";
