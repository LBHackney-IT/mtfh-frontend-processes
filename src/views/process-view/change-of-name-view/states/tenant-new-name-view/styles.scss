@import "../../../../../../node_modules/lbh-frontend/lbh/base";
@import "../../../../../../node_modules/govuk-frontend/govuk/helpers/media-queries";

.mtfh-person-form {
  @include govuk-media-query($from: tablet) {
    max-width: 458px;
  }

  &__input {
    &--full {
      width: 100%;
    }

    &--contained {
      width: 300px;
      max-width: 100%;
    }
  }
}
