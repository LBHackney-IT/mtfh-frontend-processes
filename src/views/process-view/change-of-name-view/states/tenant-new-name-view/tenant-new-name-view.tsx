import { Form, Formik } from "formik";

import {
  TenantNewNameFormData,
  tenantNewNameSchema,
} from "../../../../../schemas/change-of-name/tenant-new-name";
import { locale } from "../../../../../services";
import { Trigger } from "../../../../../services/processes/types";

import { PersonTitle } from "@mtfh/common/lib/api/person/v1";
import { Process, editProcess } from "@mtfh/common/lib/api/process/v1";
import {
  Button,
  Center,
  Field,
  FormGroup,
  Heading,
  Input,
  Select,
  Spinner,
} from "@mtfh/common/lib/components";
import { useErrorCodes } from "@mtfh/common/lib/hooks";

import "./styles.scss";

interface TenantNewNameViewProps {
  process: Process;
  mutate: () => void;
  setGlobalError: any;
}

export const TenantNewNameView = ({
  process,
  mutate,
  setGlobalError,
}: TenantNewNameViewProps) => {
  const errorMessages = useErrorCodes();

  if (!errorMessages) {
    return (
      <Center>
        <Spinner />
      </Center>
    );
  }

  return (
    <div data-testid="changeofname-EnterNewName">
      <Heading variant="h3">Enter tenant's new name</Heading>
      <Formik<TenantNewNameFormData>
        initialValues={{ title: "", firstName: "", middleName: "", surname: "" }}
        validateOnChange={false}
        validateOnBlur={false}
        validationSchema={tenantNewNameSchema(errorMessages)}
        onSubmit={async (data) => {
          try {
            await editProcess({
              id: process.id,
              processTrigger: Trigger.EnterNewName,
              processName: process?.processName,
              etag: process.etag || "",
              formData: {
                title: data.title,
                firstName: data.firstName,
                middleName: data.middleName,
                surname: data.surname,
              },
              documents: [],
            });
            mutate();
          } catch (e: any) {
            setGlobalError(e.response?.status || 500);
          }
        }}
      >
        {({ dirty, isSubmitting, validateForm }) => {
          return (
            <Form id="person-form" className="mtfh-person-form" noValidate>
              <FormGroup id="person-form-new-tenant-name" name="new-tenant-name">
                <>
                  <Field
                    id="person-form-title"
                    label={locale.views.person.title}
                    name="title"
                    required
                  >
                    <Select
                      data-testid="mtfh-person-form-title"
                      className="mtfh-person-form__input--contained"
                    >
                      <option value="">{locale.views.person.titlePlaceholder}</option>
                      {(Object.keys(PersonTitle) as Array<keyof typeof PersonTitle>).map(
                        (key) => (
                          <option key={key} value={PersonTitle[key]}>
                            {PersonTitle[key]}
                          </option>
                        ),
                      )}
                    </Select>
                  </Field>
                  <Field
                    id="person-form-firstName"
                    label={locale.views.person.firstName}
                    name="firstName"
                    required
                  >
                    <Input placeholder={locale.views.person.firstNamePlaceholder} />
                  </Field>
                  <Field
                    id="person-form-middleName"
                    name="middleName"
                    label={locale.views.person.middleName}
                  >
                    <Input placeholder={locale.views.person.middleNamePlaceholder} />
                  </Field>
                  <Field
                    id="person-form-surname"
                    name="surname"
                    label={locale.views.person.surname}
                    required
                  >
                    <Input placeholder={locale.views.person.surnamePlaceholder} />
                  </Field>
                </>
              </FormGroup>
              <div className="start-process__actions">
                <Button
                  onClick={() => validateForm()}
                  disabled={!dirty}
                  isLoading={isSubmitting}
                  loadingText={locale.loadingText}
                  type="submit"
                >
                  {locale.next}
                </Button>
              </div>
            </Form>
          );
        }}
      </Formik>
    </div>
  );
};
