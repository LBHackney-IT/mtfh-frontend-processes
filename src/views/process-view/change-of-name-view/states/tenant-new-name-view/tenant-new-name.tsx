import { useState } from "react";

import { Form, Formik } from "formik";
import * as Yup from "yup";

import { EntitySummary } from "../../../../../components";
import { locale } from "../../../../../services";
import { IProcess } from "../../../../../types";
import { areAllTruthy } from "../../../../../utils/validation";

import { PersonTitle } from "@mtfh/common/lib/api/person/v1";
import { Process, editProcess } from "@mtfh/common/lib/api/process/v1";
import {
  Button,
  ErrorSummary,
  Field,
  FormGroup,
  Heading,
  Input,
  Select,
  StatusErrorSummary,
} from "@mtfh/common/lib/components";
import "./styles.scss";

interface TenantNewNameViewProps {
  processConfig: IProcess;
  process: Process;
  mutate: () => void;
  // eslint-disable-next-line react/no-unused-prop-types
  optional?: any;
}

export const schema = Yup.object({
  title: Yup.string().required(),
  firstName: Yup.string().required(),
  middleName: Yup.string().optional(),
  surname: Yup.string().required(),
});

export type FormData = Yup.Asserts<typeof schema>;

export const TenantNewName = ({
  processConfig,
  process,
  mutate,
}: TenantNewNameViewProps) => {
  const stateConfig = processConfig.states.enterNewName;
  const error = undefined;
  const [globalError, setGlobalError] = useState<number>();

  if (error) {
    return (
      <ErrorSummary
        id="tenant-new-name-view"
        title={locale.errors.unableToFetchRecord}
        description={locale.errors.unableToFetchRecordDescription}
      />
    );
  }

  return (
    <div data-testid="changeofname-EnterNewName">
      <Heading variant="h1">{processConfig.title}</Heading>
      <EntitySummary id={process.targetId} type={processConfig.targetType} />
      {globalError && (
        <StatusErrorSummary id="select-tenant-global-error" code={globalError} />
      )}
      <Heading variant="h3">Enter tenant's new name</Heading>
      <Formik<FormData>
        initialValues={{ title: "", firstName: "", middleName: "", surname: "" }}
        validateOnChange={false}
        validateOnBlur={false}
        validationSchema={schema}
        isInitialValid={false}
        onSubmit={async (data) => {
          try {
            await editProcess({
              id: process.id,
              processTrigger: stateConfig.triggers.enterNewName,
              processName: process?.processName,
              etag: process.etag || "",
              formData: {
                title: data.title,
                firstName: data.firstName,
                middleName: data.middleName,
                surname: data.surname,
              },
              documents: [],
            });
            mutate();
          } catch (e: any) {
            setGlobalError(e.response?.status || 500);
          }
        }}
      >
        {({ isValid, values, isSubmitting, validateForm }) => {
          return (
            <Form id="person-form" className="mtfh-person-form">
              <FormGroup id="person-form-new-tenant-name" name="new-tenant-name">
                <>
                  <Field
                    id="person-form-title"
                    label={locale.views.person.title}
                    name="title"
                    required
                  >
                    <Select className="mtfh-person-form__input--contained">
                      <option value="">{locale.views.person.titlePlaceholder}</option>
                      {(Object.keys(PersonTitle) as Array<keyof typeof PersonTitle>).map(
                        (key) => (
                          <option key={key} value={PersonTitle[key]}>
                            {PersonTitle[key]}
                          </option>
                        ),
                      )}
                    </Select>
                  </Field>
                  <Field
                    id="person-form-firstName"
                    label={locale.views.person.firstName}
                    name="firstName"
                    required
                  >
                    <Input placeholder={locale.views.person.firstNamePlaceholder} />
                  </Field>
                  <Field
                    id="person-form-middleName"
                    name="middleName"
                    label={locale.views.person.middleName}
                  >
                    <Input placeholder={locale.views.person.middleNamePlaceholder} />
                  </Field>
                  <Field
                    id="person-form-surname"
                    name="surname"
                    label={locale.views.person.surname}
                    required
                  >
                    <Input placeholder={locale.views.person.surnamePlaceholder} />
                  </Field>
                </>
              </FormGroup>
              <div className="start-process__actions">
                <Button
                  onClick={() => validateForm()}
                  disabled={
                    !areAllTruthy(values.title, values.firstName, values.surname) ||
                    !isValid
                  }
                  isLoading={isSubmitting}
                  loadingText={locale.loadingText}
                  type="submit"
                >
                  {locale.next}
                </Button>
              </div>
            </Form>
          );
        }}
      </Formik>
    </div>
  );
};
