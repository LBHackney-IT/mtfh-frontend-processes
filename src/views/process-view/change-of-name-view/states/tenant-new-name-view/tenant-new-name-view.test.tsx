import React from "react";

import {
  getContactDetailsV2,
  getReferenceDataV1,
  mockContactDetailsV2,
  mockProcessV1,
  patchProcessV1,
  render,
  server,
} from "@hackney/mtfh-test-utils";
import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { changeofname } from "../../../../../services/processes";
import { Trigger } from "../../../../../services/processes/types";
import { TenantNewNameView } from "./tenant-new-name-view";

import * as processApi from "@mtfh/common/lib/api/process/v1/service";
import commonLocale from "@mtfh/common/lib/locale";

const options = {
  url: "/processes/changeofname/e63e68c7-84b0-3a48-b450-896e2c3d7735",
  path: "/processes/:processName/:processId",
};

describe("changeofname/tenant-new-name-view", () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  test("it renders TenantNewNameView correctly on state EnterNewName", async () => {
    server.use(getReferenceDataV1({}));
    server.use(getContactDetailsV2(mockContactDetailsV2));
    const { container } = render(
      <TenantNewNameView
        process={{
          ...mockProcessV1,
          currentState: {
            ...mockProcessV1.currentState,
            state: changeofname.states.enterNewName.state,
          },
        }}
        mutate={() => {}}
        setGlobalError={() => {}}
      />,
      options,
    );
    await expect(
      screen.findByTestId("changeofname-EnterNewName"),
    ).resolves.toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test("it shows validation messages on the relevant fields if they are empty", async () => {
    server.use(getReferenceDataV1({}));
    server.use(getContactDetailsV2(mockContactDetailsV2));
    const user = userEvent.setup();
    render(
      <TenantNewNameView
        process={{
          ...mockProcessV1,
          currentState: {
            ...mockProcessV1.currentState,
            state: changeofname.states.enterNewName.state,
          },
        }}
        mutate={() => {}}
        setGlobalError={() => {}}
      />,
      options,
    );

    const nextButton = await screen.findByText("Next");
    expect(nextButton).toBeDisabled();

    await user.type(screen.getByPlaceholderText("Enter first name"), "t");
    await user.type(screen.getByPlaceholderText("Enter middle name"), "t$");
    expect(nextButton).not.toBeDisabled();

    await user.click(nextButton);
    await expect(
      screen.findByText(commonLocale.hooks.defaultErrorMessages.W5),
    ).resolves.toBeInTheDocument();
    await expect(
      screen.queryByText(commonLocale.hooks.defaultErrorMessages.W56),
    ).not.toBeInTheDocument();
    await expect(
      screen.queryByText(commonLocale.hooks.defaultErrorMessages.W8),
    ).toBeInTheDocument();
    await expect(
      screen.findByText(commonLocale.hooks.defaultErrorMessages.W57),
    ).resolves.toBeInTheDocument();
  });

  test("it submits the form", async () => {
    server.use(getReferenceDataV1({}));
    server.use(getContactDetailsV2(mockContactDetailsV2));
    const editProcessSpy = jest.spyOn(processApi, "editProcess");
    const user = userEvent.setup();
    render(
      <TenantNewNameView
        process={{
          ...mockProcessV1,
          currentState: {
            ...mockProcessV1.currentState,
            state: changeofname.states.enterNewName.state,
          },
        }}
        mutate={() => {}}
        setGlobalError={() => {}}
      />,
      options,
    );

    const nextButton = await screen.findByText("Next");

    await user.selectOptions(screen.getByTestId("mtfh-person-form-title"), "Mr");
    await user.type(screen.getByPlaceholderText("Enter first name"), "name");
    await user.type(screen.getByPlaceholderText("Enter last name"), "lastname");

    await user.click(nextButton);

    expect(editProcessSpy).toBeCalledWith({
      id: mockProcessV1.id,
      processTrigger: Trigger.EnterNewName,
      processName: mockProcessV1?.processName,
      etag: "",
      formData: {
        title: "Mr",
        firstName: "name",
        middleName: "",
        surname: "lastname",
      },
      documents: [],
    });
  });

  test("it renders error when submit fails", async () => {
    server.use(patchProcessV1({}, 500));
    server.use(getReferenceDataV1({}));
    server.use(getContactDetailsV2(mockContactDetailsV2));
    render(
      <TenantNewNameView
        process={{
          ...mockProcessV1,
          currentState: {
            ...mockProcessV1.currentState,
            state: changeofname.states.enterNewName.state,
          },
        }}
        mutate={() => {}}
        setGlobalError={() => {}}
      />,
      options,
    );
    const nextButton = await screen.findByText("Next");
    await userEvent.selectOptions(screen.getByTestId("mtfh-person-form-title"), "Mr");
    await userEvent.type(screen.getByPlaceholderText("Enter first name"), "name");
    await userEvent.type(screen.getByPlaceholderText("Enter last name"), "lastname");
    await userEvent.click(nextButton);

    expect(
      screen.findByText(commonLocale.components.statusErrorSummary.statusTitle(500)),
    ).resolves.toBeInTheDocument();
  });
});
