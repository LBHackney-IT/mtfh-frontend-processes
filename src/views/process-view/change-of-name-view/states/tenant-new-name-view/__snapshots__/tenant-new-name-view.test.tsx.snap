// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changeofname/tenant-new-name-view it renders TenantNewNameView correctly on state EnterNewName 1`] = `
<div>
  <div
    data-testid="changeofname-EnterNewName"
  >
    <h3
      class="lbh-heading-h3"
    >
      Enter tenant's new name
    </h3>
    <form
      action="#"
      class="mtfh-person-form"
      id="person-form"
      novalidate=""
    >
      <div
        class="govuk-form-group lbh-form-group"
        id="person-form-new-tenant-name"
      >
        <label
          class="govuk-label lbh-label"
          for="person-form-new-tenant-name-field"
        >
          
        </label>
        <div
          class="govuk-form-group lbh-form-group"
          id="person-form-title"
        >
          <label
            class="govuk-label lbh-label"
            for="person-form-title-field"
          >
            Title
            <sup
              aria-hidden="true"
            >
              *
            </sup>
          </label>
          <select
            class="govuk-select lbh-select mtfh-person-form__input--contained"
            data-testid="mtfh-person-form-title"
            id="person-form-title-field"
            name="title"
            required=""
          >
            <option
              value=""
            >
              Select a title
            </option>
            <option
              value="Mr"
            >
              Mr
            </option>
            <option
              value="Mrs"
            >
              Mrs
            </option>
            <option
              value="Miss"
            >
              Miss
            </option>
            <option
              value="Ms"
            >
              Ms
            </option>
            <option
              value="Dr"
            >
              Dr
            </option>
            <option
              value="Master"
            >
              Master
            </option>
            <option
              value="Rabbi"
            >
              Rabbi
            </option>
            <option
              value="Reverend"
            >
              Reverend
            </option>
            <option
              value="Other"
            >
              Other
            </option>
          </select>
        </div>
        <div
          class="govuk-form-group lbh-form-group"
          id="person-form-firstName"
        >
          <label
            class="govuk-label lbh-label"
            for="person-form-firstName-field"
          >
            First name
            <sup
              aria-hidden="true"
            >
              *
            </sup>
          </label>
          <input
            class="govuk-input lbh-input"
            id="person-form-firstName-field"
            name="firstName"
            placeholder="Enter first name"
            required=""
            value=""
          />
        </div>
        <div
          class="govuk-form-group lbh-form-group"
          id="person-form-middleName"
        >
          <label
            class="govuk-label lbh-label"
            for="person-form-middleName-field"
          >
            Middle name
            
          </label>
          <input
            class="govuk-input lbh-input"
            id="person-form-middleName-field"
            name="middleName"
            placeholder="Enter middle name"
            value=""
          />
        </div>
        <div
          class="govuk-form-group lbh-form-group"
          id="person-form-surname"
        >
          <label
            class="govuk-label lbh-label"
            for="person-form-surname-field"
          >
            Last name
            <sup
              aria-hidden="true"
            >
              *
            </sup>
          </label>
          <input
            class="govuk-input lbh-input"
            id="person-form-surname-field"
            name="surname"
            placeholder="Enter last name"
            required=""
            value=""
          />
        </div>
      </div>
      <div
        class="start-process__actions"
      >
        <button
          class="govuk-button lbh-button"
          disabled=""
          type="submit"
        >
          Next
        </button>
      </div>
    </form>
  </div>
</div>
`;
